tools/chpldef/protocol-types.cpp

```diff
@@ -121,75 +124,176 @@ bool TraceLevel::fromJson(const JsonValue& j, JsonPath p) {
   return false;
 }
 
-JsonValue TraceLevel::toJson() const {
-  CHPLDEF_IMPOSSIBLE();
-  return nullptr;
+bool WorkspaceFolder::fromJson(const JsonValue& j, JsonPath p) {
+  JsonMapper m(j, p);
+  return m && MAP_(m, uri) && MAP_(m, name);
 }
 
-bool WorkspaceFolder::fromJson(const JsonValue& j, JsonPath p) {
-  CHPLDEF_TODO();
-  return true;
+JsonValue ServerInfo::toJson() const {
+  JsonObject ret;
+  FIELD_(ret, name);
+  FIELD_(ret, version);
+  return ret;
 }
 
-JsonValue WorkspaceFolder::toJson() const {
-  CHPLDEF_IMPOSSIBLE();
-  return nullptr;
+JsonValue ServerCapabilities::toJson() const {
+  JsonObject ret;
+  FIELD_(ret, positionEncoding);
+  FIELD_(ret, textDocumentSync);
+  FIELD_(ret, notebookDocumentSync);
+  FIELD_(ret, completionProvider);
+  FIELD_(ret, hoverProvider);
+  FIELD_(ret, signatureHelpProvider);
+  FIELD_(ret, declarationProvider);
+  FIELD_(ret, definitionProvider);
+  FIELD_(ret, typeDefinitionProvider);
+  FIELD_(ret, implementationProvider);
+  FIELD_(ret, referencesProvider);
+  FIELD_(ret, documentHighlightProvider);
+  FIELD_(ret, documentSymbolProvider);
+  FIELD_(ret, codeActionProvider);
+  FIELD_(ret, codeLensProvider);
+  FIELD_(ret, documentLinkProvider);
+  FIELD_(ret, colorProvider);
+  FIELD_(ret, documentFormattingProvider);
+  FIELD_(ret, documentRangeFormattingProvider);
+  FIELD_(ret, documentOnTypeFormattingProvider);
+  FIELD_(ret, renameProvider);
+  FIELD_(ret, foldingRangeProvider);
+  FIELD_(ret, executeCommandProvider);
+  FIELD_(ret, selectionRangeProvider);
+  FIELD_(ret, linkEditingRangeProvider);
+  FIELD_(ret, callHierarchyProvider);
+  FIELD_(ret, semanticTokensProvider);
+  FIELD_(ret, monikerProvider);
+  FIELD_(ret, typeHierarchyProvider);
+  FIELD_(ret, inlineValueProvider);
+  FIELD_(ret, inlayHintProvider);
+  FIELD_(ret, diagnosticProvider);
+  FIELD_(ret, workspaceSymbolProvider);
+  FIELD_(ret, workspace);
+  FIELD_(ret, experimental);
+  return ret;
 }
 
-bool ServerInfo::fromJson(const JsonValue& j, JsonPath p) {
-  CHPLDEF_IMPOSSIBLE();
-  return true;
+JsonValue TextDocumentSyncOptions::toJson() const {
+  JsonObject ret;
+  FIELD_(ret, openClose);
+  if (change) ret["change"] = static_cast<int>(*change);
+  return ret;
 }
 
-JsonValue ServerInfo::toJson() const {
-  JsonObject ret { FIELD_(name), FIELD_(version) };
+bool TextDocumentItem::fromJson(const JsonValue& j, JsonPath p) {
+  JsonMapper m(j, p);
+  bool ret = m;
+  ret &= MAP_(m, uri);
+  ret &= MAP_(m, languageId);
+  ret &= MAP_(m, version);
+  ret &= MAP_(m, text);
```

Comment:
Rewriting this using a multi-line chain of && would allow short-circuiting and possibly save some work in the badbadbad case.